File: app/src/main/java/com/bove/martin/pexel/viewmodels/MainActivityViewModel.java
package com.bove.martin.pexel.viewmodels;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.bove.martin.pexel.model.Foto;
import com.bove.martin.pexel.model.Search;
import com.bove.martin.pexel.repositories.FotosRepository;
import com.bove.martin.pexel.repositories.PopularSearchsRepository;

import java.util.List;

/**
 * Created by Martín Bove on 01-Feb-20.
 * E-mail: mbove77@gmail.com
 */
public class MainActivityViewModel extends ViewModel {

    private MutableLiveData<List<Foto>> mFotos;
    private MutableLiveData<List<Search>> mSearchs;
    private MutableLiveData<String> mQueryString;

    private FotosRepository mRepo;
    private PopularSearchsRepository mPopularSearchRepo;
    private int pageNumber = 1;

    public MainActivityViewModel(){
        mRepo = FotosRepository.getInstance();
        mPopularSearchRepo = PopularSearchsRepository.getInstance();
        mFotos = new MutableLiveData<List<Foto>>();
        mQueryString = new MutableLiveData<String>();
    }

    public void addPage() {
        pageNumber ++;
    }

    private void resetList() { pageNumber = 1; }

    public LiveData<List<Foto>> getFotos(Boolean resetList) {
        if(resetList) { resetList(); }
        return mFotos = mRepo.getFotos(mQueryString.getValue(), pageNumber, resetList);
    }

    public LiveData<List<Search>> getSearchs() {
        return mSearchs = mPopularSearchRepo.getSearchs();
    }

    public void setQueryString(String query) {
        mQueryString.postValue(query);
    }

    public LiveData<String> getQueryString() {
        return mQueryString;
    }

}

File: app/src/main/java/com/bove/martin/pexel/presentation/MainActivityViewModel.kt
package com.bove.martin.pexel.presentation

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bove.martin.pexel.di.DispacherModule.MainDispatcher
import com.bove.martin.pexel.domain.GetFotosUseCase
import com.bove.martin.pexel.domain.GetPupularSearchesUseCase
import com.bove.martin.pexel.domain.GetSearchedFotosUseCase
import com.bove.martin.pexel.domain.model.Foto
import com.bove.martin.pexel.domain.model.Search
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import javax.inject.Inject

/**
 * Created by Martín Bove on 01-Feb-20.
 * E-mail: mbove77@gmail.com
 */

@HiltViewModel
class MainActivityViewModel @Inject constructor(
    private val getFotosUseCase: GetFotosUseCase,
    private val getPopularSearchesUseCase: GetPupularSearchesUseCase,
    private val getSearchedFotosUseCase: GetSearchedFotosUseCase,
    @MainDispatcher private val mainDispatcher: CoroutineDispatcher
    ) : ViewModel() {

    val fotos = MutableLiveData<List<Foto>>()
    private val _fotos: MutableList<Foto> = arrayListOf()

    val searches = MutableLiveData<List<Search>>()

    private var queryString:String? = null
    private var pageNumber = 0


    fun getFotos() {
        pageNumber++
        if(queryString.isNullOrEmpty())
            getCuratedFotos()
        else
            getSearchedFotos()
    }

    private fun getCuratedFotos() {
        viewModelScope.launch {
            val response = getFotosUseCase(pageNumber)
            if (response.operationResult) {
                withContext(mainDispatcher) {
                    val listOfFotos = (response.resultObject as List<*>).filterIsInstance<Foto>()
                    _fotos.addAll(listOfFotos)
                    fotos.postValue(_fotos)
                }
            } else {
                fotos.postValue(null)
            }
        }
    }

    private fun getSearchedFotos() {
        viewModelScope.launch {
            val response = getSearchedFotosUseCase(queryString, pageNumber)
            if (response.operationResult) {
                withContext(mainDispatcher) {
                    val listOfFotos = (response.resultObject as List<*>).filterIsInstance<Foto>()
                    _fotos.addAll(listOfFotos)
                    fotos.postValue(_fotos)
                }
            } else {
                fotos.postValue(null)
            }
        }
    }

    fun getSearchOptions() {
        viewModelScope.launch {
            val response = getPopularSearchesUseCase()

            if (response.isNotEmpty()) {
                withContext(mainDispatcher) {
                    searches.postValue(response)
                }
            }
        }
    }

    fun setQueryString(search: String?) {
       if(!search.isNullOrEmpty()) {
           queryString = search
           resetFotoList()
       }
    }

    fun getQueryString(): String? {
        return queryString
    }

    fun clearQueryString() {
        queryString = null
        resetFotoList()
    }

    private fun resetFotoList() {
        pageNumber = 0
        _fotos.clear()
    }
}